File: Solutions/Python/word_pattern.py
"""
    [EASY]
    290. Word Pattern

    Concepts:
    - string
    - hash table

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 16.41 MB  [Beats 100%]
"""

class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        pat = list(pattern)
        string = s.split()
        p_set = set()
        s_set = set()

        if len(pat) != len(string):
            return False

        c_to_s = {}
        for p, w in zip(pat, string):
            if p not in c_to_s:
                c_to_s.update({p : w})
            else:
                if c_to_s[p] != w:
                    return False
            p_set.add(p)
            s_set.add(w)

        return len(p_set) == len(s_set)

File: Solutions/Cpp/word_pattern.cpp
/*
    [EASY]
    290. Word Pattern

    Concepts:
    - string
    - hash table

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 8.97 MB   [Beats 12.31%]
*/

class Solution {
public:
    bool wordPattern(string pattern, string str) {
        std::unordered_set<char> p_set = {};
        std::unordered_set<std::string> s_set = {};
        std::unordered_map<char, std::string> c_to_s = {};

        // split string on spaces
        std::vector<std::string> s;
        std::stringstream ss(str);
        std::string item;

        while (ss >> item)
            s.push_back(item);

        if (pattern.size() != s.size())
            return false;

        for (int i = 0; i < s.size(); i++) {
            if (!c_to_s.contains(pattern[i]))
                c_to_s[pattern[i]] = s[i];
            else if (c_to_s[pattern[i]] != s[i])
                return false;

            p_set.insert(pattern[i]);
            s_set.insert(s[i]);
        }

        return p_set.size() == s_set.size();
    }
};